lib/ClangToAst.ml

```diff
@@ -1396,6 +1482,63 @@ let rec translate_stmt (env : env) (s : Clang.Ast.stmt_desc) : Krml.Ast.expr =
     end
   | If { cond = { desc = BinaryOperator { lhs; kind = NE; rhs }; _ }; then_branch; _ }
     when has_pointer_type lhs && is_null rhs -> translate_stmt env then_branch.desc
+
+  | If { cond; then_branch; else_branch; _ } when is_tag_check env cond ->
+      let var, variant, case_ref = deconstruct_tag_check env cond in
+
+      let lid = Helpers.assert_tlid var.typ in
+      let case, fs = match LidMap.find_opt lid !type_def_map with
+        | Some (lazy (Variant branches)) -> List.nth branches variant
+        | _ -> fatal_error "Expected a tagged union expression"
+      in
+
+      let name, case_t = match fs with
+        | [(n, (t, _))] -> n, t
+        | _ -> failwith "Tagged union case has more than one field"
+      in
+
+      let binder = Helpers.fresh_binder name case_t in
+      let pat = Krml.Ast.with_type case_t (PBound 0) in
+      (* We need to add the new binder to the environment before translating
+         the branches.
+         We also need to be able to later retrieve the content of the variant
+         constructor when accessing the corresponding field.
+         To do so, we store in the environment a variable called v!!{atom},
+         where {atom} is replaced by the binder's atom.
+         We then update the environment for `var` to store that it is a tagged
+         union in case `case`, and that the corresponding content is ``v!!{atom}`.
+         Importantly, `v!!{atom}` is not a valid variable name, and therefore does
+         not conflict with existing variables.
+      *)
+      let env_binder_name = binder.node.name ^ "!!" ^ show_atom_t binder.node.atom in
```

Comment:
can we have some metadata on the variables? if you make variables a record type, you could have a `metadata: meta list` field in your variables, where `type meta = | VariantVar ...` and can be extended in the future

relying on a debug printer to encode some semantic information is likely to break in the future -- it would also help to make the code easier to maintain in the future

Reply:
I agree, I wasn't a big fan of using show for this. I'm wondering what the VariantVar meta would look like though, would it include a unique identifier, which would be what we are storing in the corresponding tagged union variable?

Reply:
To clarify, the main issue is that I need a way to retrieve the VariantVar associated to a given case, however, we currently do not have unique identifiers for variables, we only have access to the name (that can be overloaded)

Reply:
you could store a DeBruijn level (i.e. the opposite of an index) and reconstruct a de bruijn variable from that level

Reply:
so:

```
type meta = | VariantVar of level

let set_case env (scrutinee: index) (case_variable: index) =
  let case_variable_level = List.length env.vars - case_variable - 1 in
  { env with vars = List.mapi (fun i var -> if i = scrutinee then { var with meta = VariantVar case_variable_level :: var.meta }) env.vars }
```